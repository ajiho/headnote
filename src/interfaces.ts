export interface HeadnoteOptions {
  banner?: string
  exclude?: string[]
  dryRun?: boolean
}
