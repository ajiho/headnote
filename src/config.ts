import { resolve } from "node:path"
import { lilconfig } from "lilconfig"
import { access, constants } from "node:fs/promises"

const CONFIG_NAME = "headnote"
const explorer = lilconfig(CONFIG_NAME)

async function loadConfig(customPath?: string) {
  let result
  if (customPath) {
    // 传递了config选项
    const filepath = resolve(process.cwd(), customPath)

    // 检查该文件是否存在
    const exists = await fileExists(filepath)

    if (!exists) {
      throw new Error(`No ${CONFIG_NAME} configuration found.`)
    }
    result = await explorer.load(filepath)
  } else {
    result = await explorer.search()
  }

  return result?.config || {}
}

async function fileExists(filePath: string) {
  try {
    await access(filePath, constants.F_OK)
    return true
  } catch {
    return false
  }
}

export { loadConfig }
