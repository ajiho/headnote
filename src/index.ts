import { readFile, writeFile } from "node:fs/promises"
import path from "node:path"
import { glob } from "tinyglobby"

import type { HeadnoteOptions } from "./interfaces.js"
export type * from "./interfaces.js"

const DEFAULT: HeadnoteOptions = {
  banner: "",
  exclude: [],
  dryRun: false,
}
function defineConfig(config: HeadnoteOptions): HeadnoteOptions {
  return config
}
async function headnote(
  patterns: string | string[],
  options: HeadnoteOptions = {},
): Promise<void> {
  // 合并选项
  // eslint-disable-next-line prefer-const
  let { banner, exclude, dryRun } = { ...DEFAULT, ...(options ?? {}) }

  if (!banner) {
    banner = await getDefaultBanner()
  }

  const files = await glob(patterns, { ignore: exclude })

  for (const file of files) {
    const content = await readFile(file, "utf-8")

    // 避免重复插入
    if (content.startsWith(banner)) {
      console.warn(`skip: ${file}`)
      continue
    }

    const newContent = `${banner}\n${content}`

    if (dryRun) {
      console.warn(`[dry-run] ${file}`)
      continue
    }

    await writeFile(file, newContent, "utf-8")
    console.log(`update: ${file}`)
  }
}

async function getDefaultBanner() {
  try {
    const pkgPath = path.resolve(process.cwd(), "package.json")
    const pkgContent = await readFile(pkgPath, "utf-8")
    const pkg = JSON.parse(pkgContent)

    const name = pkg.name || "Unknown"
    const version = pkg.version || ""
    const homepage = pkg.homepage || ""
    const license = pkg.license || "UNLICENSED"
    const author = pkg.author || ""
    const year = new Date().getFullYear()

    return `/*!
 * ${name.charAt(0).toUpperCase() + name.slice(1)} v${version}
 * ${homepage}
 *
 * Copyright (c) ${year} ${author}
 * Licensed under the ${license} license
 */`
  } catch (err) {
    return "/* */"
  }
}

export { headnote, defineConfig }
