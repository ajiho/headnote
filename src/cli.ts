#!/usr/bin/env node
import { Command } from "commander"
import { loadConfig } from "./config.js"
import { headnote } from "./index.js"
import pkg from "../package.json" with { type: "json" }

const program = new Command()

program
  .name("headnote")
  .version(pkg.version)
  .description("Prepend banner headers to files.")
  .option("-c, --config <path>", "Path to config file")
  .option(
    "--exclude <patterns...>",
    "Exclude files or directories (supports glob)",
  )
  .option("--dry-run", "Print actions without writing files")
  .argument("<files...>", "Files or glob patterns")
  .addHelpText(
    "afterAll",
    `
Examples:
  # Prepend banner to all JS files in src/
  $ headnote "src/**/*.js"

  # Load custom config file
  $ headnote "src/**/*.js" --config headnote.config.mjs
`,
  )
  .parse(process.argv)

const opts = program.opts()
const files = program.args

try {
  const config = await loadConfig(opts.config)
  await headnote(files, {
    ...config,
    ...opts,
  })
} catch (error) {
  console.error(error.message)
  process.exit(1)
}
