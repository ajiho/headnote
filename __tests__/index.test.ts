import {
  describe,
  it,
  expect,
  beforeAll,
  beforeEach,
  afterEach,
  afterAll,
} from "vitest"
import fs from "fs-extra"
import path from "node:path"
import { headnote, defineConfig } from "../src"
import { temporaryDirectory } from "tempy"
import escapeStringRegexp from "escape-string-regexp"

let oldCwd: string
let TMP_DIR: string

describe("headnote", () => {
  let filePath: string
  let banner: string

  beforeAll(() => {
    oldCwd = process.cwd()
  })

  beforeEach(() => {
    TMP_DIR = temporaryDirectory()
    fs.ensureDirSync(TMP_DIR)
    process.chdir(TMP_DIR)

    filePath = path.join(TMP_DIR, "test.js")

    banner = "/* my banner */"
  })

  afterEach(() => {
    process.chdir(oldCwd)
    fs.removeSync(TMP_DIR)
  })

  it("应该给文件添加横幅", async () => {
    fs.outputFileSync(filePath, "console.log('hi')\n", "utf-8")

    const patterns = "test.js"

    await headnote(patterns, { banner })
    const content = fs.readFileSync(patterns, "utf-8")

    expect(content.startsWith(banner)).toBe(true)
  })

  it("应该原样返回传入的配置对象", () => {
    const input = { banner: "/* test */", exclude: ["dist"], dryRun: true }
    const result = defineConfig(input)
    expect(result).toBe(input) // 同一个引用对象
    expect(result).toEqual(input)
  })

  it("应该跳过已经设置过横幅的文件", async () => {
    fs.outputFileSync(filePath, `${banner}\nconsole.log('hi')\n`, "utf-8")

    const patterns = "test.js"
    await headnote(patterns, { banner })
    const content = fs.readFileSync(filePath, "utf-8")

    console.log(content)

    const matches =
      content.match(new RegExp(escapeStringRegexp(banner), "g")) || []

    expect(matches.length === 1).toBe(true)
  })

  it("如果没有提供banner, 如果package.json存在则从中选择字段生成横幅", async () => {
    const pkgJson = {
      name: "my-lib",
      version: "1.2.3",
      homepage: "https://example.com",
      license: "MIT",
      author: "tester",
    }

    fs.outputJSONSync("package.json", pkgJson, { spaces: 2 })
    fs.outputFileSync(filePath, "console.log('hi')\n", "utf-8")

    const patterns = "test.js"
    await headnote(patterns, {})
    const content = fs.readFileSync(filePath, "utf-8")

    expect(content.includes(`My-lib v1.2.3`)).toBe(true)
    expect(content.includes("https://example.com")).toBe(true)
  })

  it("如果没有提供banner, 如果package.json存在但是需要的字段不存在", async () => {
    const pkgJson = { type: "module" }
    fs.outputJSONSync("package.json", pkgJson, { spaces: 2 })
    fs.outputFileSync(filePath, "console.log('hi2')\n", "utf-8")

    const patterns = "test.js"
    await headnote(patterns, {})
    const content = fs.readFileSync(filePath, "utf-8")

    expect(content).toMatchSnapshot()
  })

  it("如果没有提供banner, 如果package.json不存在返回空注释", async () => {
    fs.outputFileSync(filePath, "console.log('hi')\n", "utf-8")
    const patterns = "test.js"
    await headnote(patterns, {})
    const content = fs.readFileSync(filePath, "utf-8")

    expect(content.includes("/* */")).toBe(true)
  })

  it("dryRun为true时不应写入文件", async () => {
    const fileCnt = "console.log('hi')\n"
    fs.outputFileSync(filePath, fileCnt, "utf-8")
    const patterns = "test.js"
    await headnote(patterns, { banner, dryRun: true })
    const finalContent = fs.readFileSync(filePath, "utf-8")

    expect(finalContent).toBe(fileCnt)
  })

  it("在传入 null 时应使用默认配置", async () => {
    const fn = async () => {
      // @ts-expect-error 故意传入错误的选项
      await headnote([], null)
    }

    await expect(fn()).resolves.not.toThrow()
  })
})
