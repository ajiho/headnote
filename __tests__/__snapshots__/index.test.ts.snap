// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`headnote > 如果没有提供banner, 如果package.json存在但是需要的字段不存在 1`] = `
"/*!
 * Unknown v
 * 
 *
 * Copyright (c) 2025 
 * Licensed under the UNLICENSED license
 */
console.log('hi2')
"
`;
